### PR TITLE
[sertop] Add sername program for serialization of terms

### PR DESCRIPTION
**Pull request by palmskog**
This targets the `v8.10` branch and is intended for inclusion in the 0.7.1 release of SerAPI for Coq 8.10. The sername program is used by our [Roosterize](https://github.com/EngineeringSoftware/roosterize) toolchain for suggesting Coq lemma names. The sername program will be forward-ported to the `v8.11` branch after the Roosterize toolchain is released.

cc: @pengyunie

### CODE REVIEW

**Comment by palmskog** [comment]
Note that compilation of the `v8.10` branch with sexplib 0.13.0 and dune 2.5.1 currently gives warnings like the following:
```shell
File "serapi/serapi_protocol.mli", line 293, characters 23-32:
Error (warning 3): deprecated: Sexplib.Conv.sexp_list
[since 2019-03] use [@sexp.list] instead
```
... unless one compiles with `dune build --profile=release`. @ejgallego is this something you want to address before the 0.7.1 release or can we just go ahead and release after merging this PR?

**Comment by ejgallego** [comment]
> @ejgallego is this something you want to address before the 0.7.1 release or can we just go ahead and release after merging this PR?

I'm aware of this problem, and some other problems that newer toolchains do introduce. Not sure what we should do here.

**Comment by ejgallego** [comment]
@palmskog , how does this relate to #174

**Comment by palmskog** [comment]
> @palmskog , how does this relate to #174

It's not directly related, but but both fit into an overall plan on learning and suggesting code conventions for Coq. The work on lemma name suggestions (primary use of sername) doesn't use tactic or proof state serialization at all. 

However, the extended abstract recently accepted at the Coq workshop looks at suggesting spacing, which **does** use tactic/proof data for learning, but does not use sername at all.

**Comment by ejgallego** [comment]
I see, thanks for the explanation, adjust #174 plans as you think then; I will prepare a release shortly [actually for both v8.11 and v8.10]

**Comment by palmskog** [comment]
> I see, thanks for the explanation, adjust #174 plans as you think then; I will prepare a release shortly [actually for both v8.11 and v8.10]

Ah, so you think you can forward port to v8.11 easily? That would be great. Let me know if I can help in some way, e.g., writing release notes.

**Comment by ejgallego** [comment]
> Ah, so you think you can forward port to v8.11 easily? That would be great. Let me know if I can help in some way, e.g., writing release notes.

I can give the forward point a go; I don't expect any particular trouble. I will do 8.10 first, then 8.11

File: sertop/sername.ml [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+     (*
+     let sdoc = Stm.new_doc ndoc in
+     let dir,from,exp = l,None,Some false in
+     let mp = Libnames.qualid_of_string dir in
+     let mfrom = Option.map Libnames.qualid_of_string from in
+     Flags.silently (Vernacentries.vernac_require mfrom exp) [mp];
+     sdoc
+     *)
+     let doc,sid = Stm.new_doc ndoc in
+     let sent = Printf.sprintf "Require %s." l in
```
Comment (ejgallego) [inline sertop/sername.ml:90]: Why are you not adding this to `require_libs` above?

Comment (ejgallego) [reply sertop/sername.ml:90]: You mention a problem in the commit, but with no further details given it is hard to say more.

**Comment by ejgallego** [comment]
Pushed a fix for Travis and warning on newer sexplib, let's see how it goes.

**Comment by palmskog** [comment]
> For the 0.7 branch; I'm OK merging this PR, as long as the tool documentation states that this tool won't be available in future releases and `sertop` must be used.

This is fine by me. Do you want the statement about future releases in `sername --help` or in the release notes? You can let me know or just add it for yourself.

As for the `require_libs` code comment, if we don't keep `sername` in future versions, maybe we can just leave that code as is? There was a very specific use case that motivated that change which I may have to dig quite a lot to find.

**Comment by ejgallego** [comment]
> This is fine by me. Do you want the statement about future releases in `sername --help` or in the release notes? You can let me know or just add it for yourself.

Thanks @palmskog , I can add a small note myself.

> As for the `require_libs` code comment, if we don't keep `sername` in future versions, maybe we can just leave that code as is? There was a very specific use case that motivated that change which I may have to dig quite a lot to find.

That's fine then.

Something it would be very useful to have tho is a small test case that covers all the things this tool does; this way we can be sure `Query name` does support the required functionality.

**Comment by ejgallego** [comment]
Thinking a bit about it, I propose we mark `sername` deprecated in 8.11/8.12 once can be done with Query, but don't remove it, at least until it becomes a problem for maintenance.

File: sertop/sername.ml [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+  let open Serapi_protocol in
+  let st = Stm.state_of_id ~doc sid in
+  let sigma, env = context_of_st st in
+  let info = QueryUtil.info_of_id env line in
+  let def = if body then fst info else snd info in
+  match def with
+  | [CoqConstr def_term] ->
+     let evd = Evd.from_env env in
+     let edef_term = EConstr.of_constr def_term in
+     let gdef_term = Detyping.detype Detyping.Now false Names.Id.Set.empty env evd edef_term in
```
Comment (palmskog) [inline sertop/sername.ml:128]: > Thinking a bit about it, I propose we mark `sername` deprecated in 8.11/8.12 once can be done with Query, but don't remove it, at least until it becomes a problem for maintenance.

This is also perfectly fine with me. However, unless I'm missing something, I think the main obstacle to turning `sername` into a simple query separate from Coq's API is the  use of detyping here. For example, would this best be done by adding another query function?

**Comment by ejgallego** [comment]
> This is also perfectly fine with me. However, unless I'm missing something, I think the main obstacle to turning sername into a simple query separate from Coq's API is the use of detyping here. Would this best be solved by adding another query function?

See #155 and #196 , they request precisely that.

**Comment by ejgallego** [comment]
@palmskog this is ready IMO, can you have a look to the two latest commits?

Maybe I do push a few tweaks to easy forward-porting, but no code changes themselves.

**Comment by palmskog** [comment]
@ejgallego the last two commits look good to me.

**Pull request merged**